controllers/update_status.go

```diff
@@ -624,6 +641,80 @@ func validateProcessGroup(ctx context.Context, r *FoundationDBClusterReconciler,
 	processGroupStatus.UpdateCondition(fdbv1beta2.PodFailing, failing, cluster.Status.ProcessGroups, processGroupStatus.ProcessGroupID)
 	processGroupStatus.UpdateCondition(fdbv1beta2.PodPending, false, cluster.Status.ProcessGroups, processGroupStatus.ProcessGroupID)
 
+	if !disableTaintFeature {
+		// Update taint status
+		err = updateTaintCondition(ctx, r, cluster, pod, processGroupStatus, nodeMap, logger)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// updateTaintCondition checks pod's node taint label and update pod's taint-related condition accordingly
+func updateTaintCondition(ctx context.Context, r *FoundationDBClusterReconciler, cluster *fdbv1beta2.FoundationDBCluster,
+	pod *corev1.Pod, processGroupStatus *fdbv1beta2.ProcessGroupStatus, nodeMap map[string]*corev1.Node, logger logr.Logger) error {
+	node, ok := nodeMap[pod.Spec.NodeName]
+	log.V(4).Info("Get pod's node", "Pod", pod.Name, "Pod's node name", pod.Spec.NodeName, "node map size", len(nodeMap))
+	if !ok {
+		node = &corev1.Node{}
+		err := r.Get(ctx, client.ObjectKey{Name: pod.Spec.NodeName}, node)
+		if err != nil {
+			return fmt.Errorf("get pod %s node %s fails with error :%w", pod.Name, pod.Spec.NodeName, err)
+		}
+		nodeMap[pod.Spec.NodeName] = node
+		log.V(4).Info("Set nodeMap cache", "Pod", pod.Name, "Pod's node name", pod.Spec.NodeName, "Node", node)
+	}
+
+	// Check the tainted duration and only mark the process group tainted after the configured tainted duration
+	// TODO: https://github.com/FoundationDB/fdb-kubernetes-operator/issues/1583
+	hasMatchingTaint := false
+	for _, taint := range node.Spec.Taints {
+		hasExactMatch := hasExactMatchedTaintKey(cluster.Spec.AutomationOptions.Replacements.TaintReplacementOptions, taint.Key)
+		for _, taintConfiguredKey := range cluster.Spec.AutomationOptions.Replacements.TaintReplacementOptions {
+			taintConfiguredKeyString := pointer.StringDeref(taintConfiguredKey.Key, "")
+			if taintConfiguredKeyString == "" || pointer.Int64Deref(taintConfiguredKey.DurationInSeconds, math.MinInt64) < 0 {
+				logger.Info("Cluster's TaintReplacementOption is disabled", "Key",
+					taintConfiguredKey.Key, "DurationInSeconds", taintConfiguredKey.DurationInSeconds,
+					"Pod", pod.Name)
+				continue
+			}
+			if taint.Key == "" {
+				// Node's taint is not properly set, skip the node's taint
+				logger.Info("Taint is not properly set", "Node", node.Name, "Taint", taint)
+				continue
+			}
+			if (hasExactMatch && taint.Key != taintConfiguredKeyString) ||
+				(!hasExactMatch && taintConfiguredKeyString != "*") {
+				continue
+			}
+
+			hasMatchingTaint = true
+			processGroupStatus.UpdateCondition(fdbv1beta2.NodeTaintDetected, true, cluster.Status.ProcessGroups, processGroupStatus.ProcessGroupID)
+			// Use node taint's timestamp as the NodeTaintDetected condition's starting time
+			if !taint.TimeAdded.IsZero() && taint.TimeAdded.Time.Unix() < *processGroupStatus.GetConditionTime(fdbv1beta2.NodeTaintDetected) {
```

Comment:
I thought we said that if the taint has not `TimeAdded` we will fall back to the current system time of the operator?

Reply:
The line above will use the system time for the condition.

This condition says "If taint _has_ TimeAdded and if it's smaller than the current time, we will use taint node's time"